gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,302 @@
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.thrift.TException;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.Striped;
+
+import gobblin.configuration.State;
+import gobblin.util.ExecutorsUtils;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * A class for registering Hive tables or partitions.
+ *
+ * An instance of this class is initialized with a {@link State} object, a {@link HiveRegistrationPolicy} and a
+ * {@link HiveStorageDescriptorManager}. The {@link #register(HiveRegistrable)} method is asynchronous and
+ * returns immediately.
+ *
+ * @author ziliu
+ */
+@Slf4j
+public class HiveRegister implements Closeable {
+
+  public static final String HIVE_DB_ROOT_DIR = "hive.db.root.dir";
+  public static final String HIVE_OWNER = "hive.owner";
+  public static final String HIVE_REGISTER_THREADS = "hive.register.threads";
+  public static final int DEFAULT_HIVE_REGISTER_THREADS = 20;
+
+  private static final String HIVE_DB_EXTENSION = ".db";
+
+  private final State state;
+  private final String hiveDbRootDir;
+  private final String hiveOwner;
+  private final HiveRegistrationPolicy partitionManager;
+  private final HiveStorageDescriptorManager sdManager;
+  private final HiveMetaStoreClient client;
+
+  private final ExecutorService executor;
+  private final List<Future<Void>> futures = Lists.newArrayList();
+  private final Striped<Lock> locks = Striped.lazyWeakLock(Integer.MAX_VALUE);
+
+  public HiveRegister(State state, HiveRegistrationPolicy partitionManager, HiveStorageDescriptorManager sdManager)
+      throws MetaException {
+    Preconditions.checkArgument(state.contains(HIVE_DB_ROOT_DIR), "Missing required property " + HIVE_DB_ROOT_DIR);
+    Preconditions.checkArgument(state.contains(HIVE_OWNER), "Missing required property " + HIVE_OWNER);
+
+    this.state = state;
+    this.hiveDbRootDir = this.state.getProp(HIVE_DB_ROOT_DIR);
+    this.hiveOwner = this.state.getProp(HIVE_OWNER);
+    this.partitionManager = partitionManager;
+    this.sdManager = sdManager;
+    this.client = new HiveMetaStoreClient(new HiveConf());
+    this.executor =
+        Executors.newFixedThreadPool(this.state.getPropAsInt(HIVE_REGISTER_THREADS, DEFAULT_HIVE_REGISTER_THREADS),
+            ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("HiveRegister")));
+  }
+
+  /**
+   * Register a {@link HiveRegistrable}, which can be a table or a partition.
+   *
+   * This method is asynchronous and returns immediately.
+   */
+  public void register(final HiveRegistrable registrable) throws IOException {
```

Comment:
This should return a ListenableFuture. Otherwise, there is no way to know if the registration succeeded or not.